contracts/account/extensions/ERC7579Modules/SocialRecoveryExecutor.sol

```diff
@@ -0,0 +1,475 @@
+        // create single call mode
+        Mode mode = ERC7579Utils.encodeMode(
+            ERC7579Utils.CALLTYPE_SINGLE, // enforce single call
+            ERC7579Utils.EXECTYPE_DEFAULT, // enforce revert on failure
+            ModeSelector.wrap(0), // no selector needed
+            ModePayload.wrap(bytes22(0)) // no payload needed
+        );
+
+        // increment nonce.
+        _recoveryConfigs[account].nonce++;
+
+        // reset recovery status.
+        _recoveryConfigs[account].recoveryStart = 0;
+
+        // execute the recovery.
+        AccountERC7579(payable(account)).executeFromExecutor(Mode.unwrap(mode), executionCalldata);
+
+        emit RecoveryExecuted(account, targetValidatorModule);
+    }
+
+    /// @notice Cancels an ongoing recovery process
+    /// @dev Can only be called by the ERC-7579 Account itself
+    function cancelRecovery() external whenRecoveryIsTimelockOrReady(msg.sender) {
+        address account = msg.sender;
+
+        _recoveryConfigs[account].recoveryStart = 0;
+
+        emit RecoveryCancelled(account);
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                SIGNATURE VERIFICATION
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @notice Computes the EIP-712 digest (typed data hash) for a recovery request
+    /// @dev Uses the current nonce to prevent replay attacks across recovery attempts
+    /// @param account The account address for which the recovery digest is being generated
+    /// @return bytes32 The EIP-712 compliant digest
+    function _getRecoveryDigest(address account) internal view returns (bytes32) {
+        bytes32 structHash = keccak256(abi.encode(RECOVERY_TYPEHASH, account, getNonce(account)));
+        return _hashTypedDataV4(structHash);
+    }
+
+    /// @notice Verifies multiple guardian signatures
+    /// @dev Checks each signature against the current recovery digest
+    /// @param account The account the signatures are for
+    /// @param guardianSignatures Array of guardian signatures to verify
+    /// @return True if all signatures are valid, false otherwise
+    function guardianSignaturesAreValid(
+        address account,
+        GuardianSignature[] calldata guardianSignatures
+    ) public view returns (bool) {
+        if (guardianSignatures.length > 32) {
+            revert MaxGuardians();
+        }
+
+        for (uint256 i = 0; i < guardianSignatures.length; i++) {
+            // check signature validity
+            if (!guardianSignatureIsValid(account, guardianSignatures[i])) {
+                return false;
+            }
+            // check for signature duplication @TBD optimize O(n^2)
+            address currentSigner = guardianSignatures[i].signer;
+            for (uint256 j = 0; j < i; j++) {
+                if (guardianSignatures[j].signer == currentSigner) {
+                    return false;
+                }
+            }
```

Comment:
This is O(n²), which is bad.

How would you make that O(n) ?

Reply:
having the signature provided in order or signer (and just checking the ordering) is what I was thinking. IMO its the simplest and most effective way. AFAIK that is what most Multisig do.